### DISCUSSION TITLE
Integration with Autogen / Comparison with Guidance

### DISCUSSION THREAD

**Discussion by shippy** [567-labs/instructor#536] [General]
Hi all! I'm currently working on constraining the output of [Autogen agents](https://github.com/microsoft/autogen) using [Guidance](https://github.com/guidance-ai/guidance). I'd like to figure out how to use Instructor instead of Guidance.

Based on [the sample Autogen/Guidance notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_guidance.ipynb), Guidance appears to link directly into the next-token generation process (?) and apply its constraints there, but only after the first-pass LLM response has been generated (? - I'm actually not 100% sure about what's going on there) and evaluated.

In contrast, my understanding of Instructor is that it tries to JSONify the LLM response, feeds it into the Pydantic model, then either succeeds or retries by feeding original prompt + whatever validation errors back to the LLM. In principle, that too could be fed into the [`register_reply` override](https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent/#register_reply) - would that be the suggested method?

(I should note that while this seems to be an Autogen/Guidance-specific question and so should be asked in other fora, there's a lack of materials on integrating almost _any_ output-constraining libraries into Autogen - and also, this is a question that's a little bit about the inner workings of Instructor. In any case, my apologies if I'm off-topic.)

**Comment by ryan-blunden** [comment]
I'm also curious about this. All of the boilerplate to create a Guidance agent in Autogen is a lot less elegant than using Instructor.

**Comment by jxnl** [comment]
This isn't something I'm planning right now, but I definitely think it makes sense.

**Comment by scruffynerf** [comment]
see https://github.com/jxnl/instructor/discussions/741
for my attempt to do an integration of Autogen/Instructor (different from above method though) 

feedback welcomed.
